File: src/MultiWall_vb/MultiWall/MonitorSelectForm.vb
Partial Public Class MonitorSelectForm
    Inherits Form
    Public Event WallpaperSelectedEvent As WallpaperSelectedEventHandler
    Private _myScreen As Screen
    Private _wallpaperFilename As String
    Private _screenIndex As Integer

    Public Sub New(ByVal whichScreen As Screen, ByVal screenIndex As Integer)
        InitializeComponent()

        _myScreen = whichScreen
        _screenIndex = screenIndex
    End Sub

    Public ReadOnly Property MyScreen() As Screen
        Get
            Return _myScreen
        End Get
    End Property

    Public ReadOnly Property WallpaperFilename() As String
        Get
            Return _wallpaperFilename
        End Get
    End Property

    Public ReadOnly Property ScreenIndex() As Integer
        Get
            Return _screenIndex
        End Get
    End Property

    Private Sub CenterOnScreen()
        Dim bounds As Rectangle = _myScreen.WorkingArea
        Me.Left = bounds.Left + ((bounds.Width - Me.Width) / 2)
        Me.Top = bounds.Top
    End Sub

    Private Sub MonitorSelectForm_DragEnter(ByVal sender As Object, ByVal e As DragEventArgs) Handles Me.DragEnter
        ' Only accept file drops
        If e.Data.GetDataPresent(DataFormats.FileDrop, False) = True Then
            e.Effect = DragDropEffects.All
        End If
    End Sub

    Private Sub MonitorSelectForm_DragDrop(ByVal sender As Object, ByVal e As DragEventArgs) Handles Me.DragDrop
        ' It's an array, even for a single file
        Dim files As String() = DirectCast(e.Data.GetData(DataFormats.FileDrop), String())

        _wallpaperFilename = files(0)
        RaiseWallpaperSelectedEvent(files(0))
    End Sub

    Private Sub RaiseWallpaperSelectedEvent(ByVal filename As String)
        Dim e As New WallpaperSelectedEventArgs()
        e.SelectedScreen = _myScreen
        e.WallpaperFilename = filename
        e.ScreenIndex = _screenIndex

        RaiseEvent WallpaperSelectedEvent(Me, e)
    End Sub

    Private Sub MonitorSelectForm_Shown(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Shown
        CenterOnScreen()
    End Sub
End Class

Public Delegate Sub WallpaperSelectedEventHandler(ByVal sender As Object, ByVal e As WallpaperSelectedEventArgs)

File: src/MultiWall_cs/MultiWall/MonitorSelectForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace MultiWall
{
    public partial class MonitorSelectForm : Form
    {
        public event WallpaperSelectedEventHandler WallpaperSelectedEvent;
        private Screen _myScreen;
        private string _wallpaperFilename;
        private int _screenIndex;

        public MonitorSelectForm(Screen whichScreen, int screenIndex)
        {
            InitializeComponent();

            _myScreen = whichScreen;
            _screenIndex = screenIndex;
        }

        public Screen MyScreen
        {
            get
            {
                return _myScreen;
            }
        }

        public string WallpaperFilename
        {
            get
            {
                return _wallpaperFilename;
            }
        }

        public int ScreenIndex
        {
            get
            {
                return _screenIndex;
            }
        }

        private void CenterOnScreen()
        {
            Rectangle bounds = _myScreen.WorkingArea;
            this.Left = bounds.Left + ((bounds.Width - this.Width) / 2);
            this.Top = bounds.Top;
        }

        private void MonitorSelectForm_DragEnter(object sender, DragEventArgs e)
        {
            // Only accept file drops
            if (e.Data.GetDataPresent(DataFormats.FileDrop, false) == true)
                e.Effect = DragDropEffects.All;
        }

        private void MonitorSelectForm_DragDrop(object sender, DragEventArgs e)
        {
            // It's an array, even for a single file
            string[] files = (string[])e.Data.GetData(DataFormats.FileDrop);

            _wallpaperFilename = files[0];
            RaiseWallpaperSelectedEvent(files[0]);
        }

        private void RaiseWallpaperSelectedEvent(string filename)
        {
            if (WallpaperSelectedEvent != null)
            {
                WallpaperSelectedEventArgs e = new WallpaperSelectedEventArgs();
                e.SelectedScreen = _myScreen;
                e.WallpaperFilename = filename;
                e.ScreenIndex = _screenIndex;

                WallpaperSelectedEvent(this, e);
            }
        }

        private void MonitorSelectForm_Shown(object sender, EventArgs e)
        {
            CenterOnScreen();
        }
    }

    public delegate void WallpaperSelectedEventHandler(object sender, WallpaperSelectedEventArgs e);
}